### ISSUE TITLE
19/20 Sem2 Q5

### ISSUE THREAD

**Issue by darylyeejh** [nus-cs2030/2021-s2#405]
I think 1 is guaranteed to be printed out because of join(), but does join() also guarantee that 1 will be printed out before the other numbers are printed out? If yes/ no then why

![Picture1](https://user-images.githubusercontent.com/79048085/116296882-4c968800-a7cd-11eb-82cb-832b1abac7ac.png)

**Comment by KingRyo13** [comment]
Yeah I think so. If I'm not wrong `join()` is blocking. So the ans for this question has to start with 1.

**Comment by JoelToh404** [comment]
yes all the answers have to start with a 1

**Comment by bandytan** [comment]
.join() waits for 1 to be completed and prints 1 out first. Since there is no .join() on 2,3,4, they might not be completed and printed out. If they are printed, 4 has to be after 2 or 3. The order of 2 and 3 does not matter.

**Comment by brendancjz** [comment]
So what should the answer for this qn be? (a), (e), (i), (j), (n)?

**Comment by Buwoo** [comment]
Fyi this problem has been discussed on #333

**Comment by rljw** [comment]
> So what should the answer for this qn be? (a), (e), (i), (j), (n)?

I think the answer for this should be (a), (e) and (n). Options (i) and (j) are not possible because` .thenRun(() -> printAsync(4)) `comes after the `.allOf()` function, meaning that 2 and 3 both have to be printed before 4 would be printed. Hope that helps!

**Comment by danieltwh** [comment]
> > So what should the answer for this qn be? (a), (e), (i), (j), (n)?
> 
> I think the answer for this should be (a), (e) and (n). Options (i) and (j) are not possible because`.thenRun(() -> printAsync(4))`comes after the `.allOf()` function, meaning that 2 and 3 both have to be printed before 4 would be printed. Hope that helps!

Yeap agreed. In this case, 1 is definitely printed. However, 2 or 3 or both could be printed after 1. If 4 is printed, then both 2 and 3 (in any order) has to be printed. Hence, the output should be (a), (e) and (n)

**Comment by chelleadel** [comment]
Hi! I agree with the others that the method .join() waits for the internal process to complete its execution before allowing any other programs to continue on. Therefore, it is a guarantee that 1 will come out first and the answer is A E I J N.

**Comment by clementyee303** [comment]
> 
> 
> Hi! I agree with the others that the method .join() waits for the internal process to complete its execution before allowing any other programs to continue on. Therefore, it is a guarantee that 1 will come out first and the answer is A E I J N.

I and J are not possible. For 4 to output, both 2 and 3 has to output too.

**Comment by NIL-99** [comment]
since 1 has a .join(), 1 will definetly be exeuted and completed first -> 1 will surely bet he first output
however, 2 and 3 are executed without a join and it runs for undetermined amount of time, hence they could either both be completed, or either being completed or none at all.
Lastly, 4 has a .thenRun(), hence can only be executed when both 2 and 3 are completed.
Therefore, the output has to have 1 being the first, 2 and 3 can be in random order or even none and lastly, should there be a 4, then both 2 and 3 have to be present and placed before 4
